src/sentry/utils/eventuser.py

```diff
@@ -96,59 +104,160 @@ def for_projects(
+                                )
+                                for column in snuba_column
+                            ],
+                        )
                     )
+            else:
+                keyword_where_conditions.append(Condition(Column(snuba_column), Op.IN, value))
+
+        if len(keyword_where_conditions) > 1:
+            where_conditions.append(
+                BooleanCondition(
+                    BooleanOp.AND if filter_boolean == "AND" else BooleanOp.OR,
+                    keyword_where_conditions,
                 )
+            )
 
-            else:
-                where_conditions.append(Condition(Column(snuba_column), Op.EQ, value))
+        if len(keyword_where_conditions) == 1:
+            where_conditions.extend(
+                keyword_where_conditions,
+            )
+
+        columns = [
+            Column("project_id"),
+            Column("group_id"),
+            Column("ip_address_v6"),
+            Column("ip_address_v4"),
+            Column("user_id"),
+            Column("user"),
+            Column("user_name"),
+            Column("user_email"),
+        ]
 
         query = Query(
             match=Entity(EntityKey.Events.value),
             select=[
-                Column("project_id"),
-                Column("group_id"),
-                Column("ip_address_v6"),
-                Column("ip_address_v4"),
-                Column("event_id"),
-                Column("user_id"),
-                Column("user"),
-                Column("user_name"),
-                Column("user_email"),
+                *columns,
+                Function("max", [Column("timestamp")], "latest_timestamp"),
             ],
             where=where_conditions,
-            limit=Limit(1),
-            orderby=[OrderBy(Column("timestamp"), Direction.DESC)],
+            groupby=[*columns],
+            orderby=[OrderBy(Column("latest_timestamp"), Direction.DESC)],
         )
 
+        if not return_all:
+            query.set_limit(1)
+
         request = Request(
             dataset=Dataset.Events.value,
             app_id=REFERRER,
             query=query,
             tenant_ids={"referrer": REFERRER, "organization_id": projects[0].organization.id},
         )
         data_results = raw_snql_query(request, referrer=REFERRER)["data"]
-        results = [EventUser.from_snuba(result) for result in data_results]
+
+        # Return the first matching item from the Snuba results.
+        # All other rows are not needed.
+        # The Snuba results are sorted by descending time.
+        first_matching_items = []
+
+        snuba_keyword_filters = {}
+        for keyword, value in keyword_filters.items():
+            snuba_column = SNUBA_KEYWORD_MAP.get_key(keyword)
+            snuba_keyword_filters[snuba_column] = value
+
+        matches = self._find_matching_items(data_results, snuba_keyword_filters, filter_boolean)
+        first_matching_items.extend(matches.values())
```

Comment:
Should we just assign directly here? Or just use `matches.values()` directly to get `results`?